BUOY: minor fixes to perception

kev-the-dev opened the pull request:
* Went through all bags I have of buoys to refine threshold values so all 3 buoys are correctly masked in each all bags
* Added circle-ness test before selecting a buoy contour, so random large noise contours should be excluded
* Made min contour area a proportion of image area so it is resistant to resolution changes.
* EDIT: update, integrated new Threshold and CircleFinder classes to replace some of the code

kev-the-dev committed:
BUOY FINDER: check circle-ness of contours, make min area proportional

kev-the-dev committed:
BUOY FINDER: tune thresholds from spreadsheet

DSsoto commented on command/sub8_launch/config/buoy_finder.yaml:
```diff
@@ -1,9 +1,10 @@
+min_observations  : 8 # Observations needed to attempt a least squares position estimate
+min_contour_area  : 0.002 # Minimum of image area contour must be to not be rejected
+max_velocity      : 1 # Highest velocity sub can be moving as in m/s before observations are ignored
+max_circle_error  : 0.4 #max 0-1 error in contour area vs enclosing circle area
```
How does this do when there are bright spots on the surface that remove parts of the contour?

kev-the-dev replied:
It accepts a pretty big error, so it still works in these cases with the bags I tested with. It more just prevents the extreme case of a giant blob that looks nothing like a circle, like the walls of ground, being the largest thresholded contour.

But yes, it might need to be tuned for these cases.

DSsoto commented on perception/sub8_perception/nodes/buoy_finder.py:
```diff
@@ -255,9 +259,10 @@ def image_cb(self, image):
         # Crop out some of the top and bottom to exclude the floor and surface reflections
         height = image.shape[0]
         roi_y = int(0.2 * height)
```
Did you choose these roi constants for operating at a particular depth?

kev-the-dev replied:
They're tuned to the bags I have to remove the worst noise in the bottom and top of frame. I changed them to ros params so they could be easily disabled or tuned if needed now.

DSsoto commented on perception/sub8_perception/nodes/buoy_finder.py:
```diff
@@ -255,9 +259,10 @@ def image_cb(self, image):
+        roi_height = height - int(0.1 * height)
         self.roi = (0, roi_y, roi_height, image.shape[1])
         self.last_image = image[self.roi[1]:self.roi[2], self.roi[0]:self.roi[3]]
+        self.image_area = self.last_image.shape[0] * self.last_image.shape[1]
```
Use `self.last_image.size`

kev-the-dev replied:
Well that would be 3 times what I want because it is a 3 channel image.

DSsoto replied:
Well, it can be used to get the right answer is all I'm saying. haha

DSsoto commented on perception/sub8_perception/nodes/buoy_finder.py:
```diff
@@ -297,15 +302,18 @@ def get_best_contour(self, contours):
         TODO: Use smarter contour filtering methods, like checking this it is circle like
```
typo? if?

DSsoto commented on perception/sub8_perception/nodes/buoy_finder.py:
```diff
@@ -297,15 +302,18 @@ def get_best_contour(self, contours):
-            tpl_center = (int(cx), int(cy))
-            return cnt, tpl_center, area
+            sort = sorted(contours, key=cv2.contourArea, reverse=True)
+            for cnt in sort:
```
Can you add a little header comment for this loop? Also, if your measure of contour changes, it might not be so good to have the return statement inside of the loop.

DSsoto reviewed (approved):
Minor comments

kev-the-dev committed:
BUOY FINDER: move roi constants to ROS params

kev-the-dev committed:
BUOY FINDER: use mil_vision's Threshold and CircleFinder to replace some code

kev-the-dev commented:
Addressed comments and replaced some code with stuff from mil_common. Just bumped in SIM again so no regressions. Should be more robust "circle" test using Hu invariants.
